### ISSUE TITLE
backward hook function

### ISSUE THREAD

**Issue by astorfi** [facebookresearch/pytorch-dp#9]
Hi,

In the [autograd_grad_sample](https://github.com/facebookresearch/pytorch-dp/blob/master/torchdp/autograd_grad_sample.py) function, you modified arguments given to the hook in-place. 

I think this should be avoided as it is related to a [known bug in PyTorch](https://github.com/pytorch/pytorch/issues/598) and may cause issues.
Is there a reason behind in-place modification instead of returning new values from the hook?

Thanks

**Comment by astorfi** [comment]
Why not simply using [clip_grad_norm_](https://pytorch.org/docs/stable/_modules/torch/nn/utils/clip_grad.html). As you mentioned in [your code here](https://github.com/facebookresearch/pytorch-dp/blob/master/torchdp/autograd_grad_sample.py), you used what suggested [here](https://github.com/cybertronai/autograd-hacks). Although it is elegant, seems to be overcomplicated.

**Comment by ilyamironov** [comment]
> Is this due to the reason that we should do the clipping before averaging the gradients in the privacy-preserving setting?

You are correct. To bound sensitivity of the SGD computation, we need to clip the L2 norm of _individual_  (per-example) gradients. The current autograd interface does not support this functionality directly, which is why we chose to use the hooks. Does it answer your question?

**Comment by astorfi** [comment]
@ilyamironov Yes. Thank you so much for your answer.

**Comment by karthikprasad** [comment]
Hi @astorfi, could you please clarify on what you mean by "in-place modification"? Which lines are you referring to?

**Comment by astorfi** [comment]
Thank you for your attention. I meant [these lines](https://github.com/facebookresearch/pytorch-dp/blob/master/torchdp/autograd_grad_sample.py#L44-L45). And the issue of in-place modification is explained [here](https://github.com/pytorch/pytorch/issues/598) in details.

**Comment by karthikprasad** [comment]
@astorfi , those lines only capture the per-sample gradients during backprop. This is needed to clip the gradients and then set the gradients of the parameters.
The issue you linked to talks about the bug in the hook capturing method when it comes to sequence of modules. This is a known issue, but unfortunately also the only known way to achieve the behaviour. This works for all the layers we support so far; perhaps we can also add a link to the issue in the README to make it clear
Most importantly, I notice that you have suggestion on a way to overcome this, we are all ears and would appreciate your contribution to the codebase.

**Comment by astorfi** [comment]
@karthikprasad Thank you for your attention. All that makes sense. Please close this issue if you believe there is no need for any more clarification from my side.
